### ISSUE TITLE
Training never starts - stuck after Loss is intialized

### ISSUE THREAD

**Issue by datamancerai** [pytorch/torchtune#2297]
I am attempting to do a full fine tune of phi3.  The model initializes but never starts training. I checked the GPU and the model is memory but there is no further utilization. 

What could be causing it to get stuck at this point?
```
tune run full_finetune_single_device --config full_ft_phi.yaml
INFO:torchtune.utils._logging:Running FullFinetuneRecipeSingleDevice with resolved config:

batch_size: 64
checkpointer:
  _component_: torchtune.training.FullModelHFCheckpointer
  checkpoint_dir: /home/LLMs/Phi-3-mini-4k-instruct
  checkpoint_files:
  - model-00001-of-00002.safetensors
  - model-00002-of-00002.safetensors
  model_type: PHI3_MINI
  output_dir: /home/LLMs/torchtune/phi3_mini/full_low_memory
  recipe_checkpoint: recipe_state.pt
compile: true
dataset:
  _component_: torchtune.datasets.instruct_dataset
  data_files: /nas/home/u315909/ctc/dummy_data_prompt.csv
  source: csv
  split: train
device: cuda
dtype: bf16
enable_activation_checkpointing: false
enable_activation_offloading: false
epochs: 8
gradient_accumulation_steps: 1
log_every_n_steps: 1
log_peak_memory_stats: true
loss:
  _component_: torchtune.modules.loss.CEWithChunkedOutputLoss
max_steps_per_epoch: null
metric_logger:
  _component_: torchtune.training.metric_logging.DiskLogger
  log_dir: /home/LLMs/torchtune/phi3_mini/full_low_memory/logs
model:
  _component_: torchtune.models.phi3.phi3_mini
optimizer:
  _component_: bitsandbytes.optim.PagedAdamW
  lr: 5.0e-06
optimizer_in_bwd: true
output_dir: /home/LLMs/torchtune/phi3_mini/full_low_memory
profiler:
  _component_: torchtune.training.setup_torch_profiler
  active_steps: 2
  cpu: true
  cuda: true
  enabled: false
  num_cycles: 1
  output_dir: /home/LLMs/torchtune/phi3_mini/full_low_memory/profiling_outputs
  profile_memory: false
  record_shapes: true
  wait_steps: 5
  warmup_steps: 3
  with_flops: false
  with_stack: false
resume_from_checkpoint: false
seed: 315909
shuffle: true
tokenizer:
  _component_: torchtune.models.phi3.phi3_mini_tokenizer
  max_seq_len: null
  path: /home/LLMs/Phi-3-mini-4k-instruct/tokenizer.model

DEBUG:torchtune.utils._logging:Setting manual seed to local seed 315909. Local seed is seed + rank = 315909 + 0
Writing logs to /home/LLMs/torchtune/phi3_mini/full_low_memory/logs/log_1737730488.txt
INFO:torchtune.utils._logging:Converting Phi-3 Mini weights from HF format.Note that conversion of adapter weights into PEFT format is not supported.
INFO:torchtune.utils._logging:Compiling model layers with torch.compile...
INFO:torchtune.utils._logging:Model is initialized with precision torch.bfloat16.
INFO:torchtune.utils._logging:Memory stats after model init:
        GPU peak memory allocation: 7.43 GiB
        GPU peak memory reserved: 7.45 GiB
        GPU peak memory active: 7.43 GiB
INFO:torchtune.utils._logging:Tokenizer is initialized from file.
INFO:torchtune.utils._logging:In-backward optimizers are set up.
INFO:torchtune.utils._logging:Compiling loss with torch.compile...
INFO:torchtune.utils._logging:Loss is initialized.
 tune run full_finetune_single_device --config full_ft_phi.yaml
INFO:torchtune.utils._logging:Running FullFinetuneRecipeSingleDevice with resolved config:

batch_size: 64
checkpointer:
  _component_: torchtune.training.FullModelHFCheckpointer
  checkpoint_dir: /home/LLMs/Phi-3-mini-4k-instruct
  checkpoint_files:
  - model-00001-of-00002.safetensors
  - model-00002-of-00002.safetensors
  model_type: PHI3_MINI
  output_dir: /home/LLMs/torchtune/phi3_mini/full_low_memory
  recipe_checkpoint: recipe_state.pt
compile: true
dataset:
  _component_: torchtune.datasets.instruct_dataset
  data_files: /nas/home/u315909/ctc/dummy_data_prompt.csv
  source: csv
  split: train
device: cuda
dtype: bf16
enable_activation_checkpointing: false
enable_activation_offloading: false
epochs: 8
gradient_accumulation_steps: 1
log_every_n_steps: 1
log_peak_memory_stats: true
loss:
  _component_: torchtune.modules.loss.CEWithChunkedOutputLoss
max_steps_per_epoch: null
metric_logger:
  _component_: torchtune.training.metric_logging.DiskLogger
  log_dir: /home/LLMs/torchtune/phi3_mini/full_low_memory/logs
model:
  _component_: torchtune.models.phi3.phi3_mini
optimizer:
  _component_: bitsandbytes.optim.PagedAdamW
  lr: 5.0e-06
optimizer_in_bwd: true
output_dir: /home/LLMs/torchtune/phi3_mini/full_low_memory
profiler:
  _component_: torchtune.training.setup_torch_profiler
  active_steps: 2
  cpu: true
  cuda: true
  enabled: false
  num_cycles: 1
  output_dir: /home/LLMs/torchtune/phi3_mini/full_low_memory/profiling_outputs
  profile_memory: false
  record_shapes: true
  wait_steps: 5
  warmup_steps: 3
  with_flops: false
  with_stack: false
resume_from_checkpoint: false
seed: 315909
shuffle: true
tokenizer:
  _component_: torchtune.models.phi3.phi3_mini_tokenizer
  max_seq_len: null
  path: /home/LLMs/Phi-3-mini-4k-instruct/tokenizer.model

DEBUG:torchtune.utils._logging:Setting manual seed to local seed 315909. Local seed is seed + rank = 315909 + 0
Writing logs to /home/LLMs/torchtune/phi3_mini/full_low_memory/logs/log_1737730488.txt
INFO:torchtune.utils._logging:Converting Phi-3 Mini weights from HF format.Note that conversion of adapter weights into PEFT format is not supported.
INFO:torchtune.utils._logging:Compiling model layers with torch.compile...
INFO:torchtune.utils._logging:Model is initialized with precision torch.bfloat16.
INFO:torchtune.utils._logging:Memory stats after model init:
        GPU peak memory allocation: 7.43 GiB
        GPU peak memory reserved: 7.45 GiB
        GPU peak memory active: 7.43 GiB
INFO:torchtune.utils._logging:Tokenizer is initialized from file.
INFO:torchtune.utils._logging:In-backward optimizers are set up.
INFO:torchtune.utils._logging:Compiling loss with torch.compile...
INFO:torchtune.utils._logging:Loss is initialized.
```

**Comment by ebsmothers** [comment]
Hi @datamancerai thanks for creating the issue. Can you try with compile=False to verify whether it’s related to that?

**Comment by datamancerai** [comment]
I’ve tried compile=False as well as fp32 with no difference.  I don’t suspect it’s related to the phi model but we are downloading another one to rule that out.

**Comment by datamancerai** [comment]
Also, the device not have external internet access in case that is relevant.

**Comment by felipemello1** [comment]
batch_size=64 and activation_checkpointing=False should likely be OOMing if using a single gpu, but that wouldnt explain it getting stuck.

as a sanity check, i would recommend trying to run one of our configs, without any modifications, e.g. llama 3.2 1b single device.

**Comment by datamancerai** [comment]
It’s a 48gb gpu. I originally had checkpointing on but turned it off while troubleshooting. I am limited in what models I can put on the machine but I will try and test a default recipe. This recipe and dataset have been used on other machines though.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by ebsmothers** [comment]
Yeah just looking through where things are logged in the recipe, it seems that if you had hit the train method you would have immediately seen [this log line](https://github.com/pytorch/torchtune/blob/5764650ec0d8472a6988784c599d67e43f31564c/recipes/full_finetune_single_device.py#L655-L657) with compile enabled. Given the logging about loss initialization is the last one you see, that'd mean something is going wrong in [this block](https://github.com/pytorch/torchtune/blob/5764650ec0d8472a6988784c599d67e43f31564c/recipes/full_finetune_single_device.py#L300-L341). Seems a bit unexpected to me, is the dataset perhaps really large or something?

**Comment by datamancerai** [comment]
I added some log statements and it is getting stuck on this line:
[https://github.com/pytorch/torchtune/blob/5764650ec0d8472a6988784c599d67e43f31564c/recipes/full_finetune_single_device.py#L563](line)

The dataset is not large, 10k rows, but I reduced it to 100 for troubleshooting.

**Comment by felipemello1** [comment]
Nice debugging @datamancerai ! Do you think you could do:

```
git clone <torchtune repo>
cd <your torchtune folder>
 "pip install -e ."
```

this will install torchtune based on this folder. Then you can go to the code in torchtune [(datasets.instruct_dataset)](https://github.com/pytorch/torchtune/blob/5764650ec0d8472a6988784c599d67e43f31564c/torchtune/datasets/_instruct.py#L15) and keep adding print statements until you find the actual culprit.

This is probably the fastest way to unblock yourself, since it seems to be specific to the machine that you are on.

**Comment by datamancerai** [comment]
@felipemello1 Thank you for pointing me in the right direction.

I've traced it back and the issue appears to be with the load_dataset call to  huggingface datasets module. I'll submit the issue with them.

**Comment by datamancerai** [comment]
I found the root cause of this. HF datasets was attempting to lock the csv file using the filelock package. The linux server did not have locks available, and the filelock package was not set up to output this error, so it was an infinite loop with no output. Once we resolved the lock availability, everything started working.
